utils/tokenwise_ablation.py

```diff
@@ -422,6 +422,269 @@ def compute_ablation_modified_metric(
+    """
+    assert cached_means is not None or direction_vectors is not None
+    file = ResultsFile(
+        "ablation_modified_loss",
+        model_name=model.cfg.model_name,
+        dataset=data_loader.name,
+        layers_to_ablate=layers_to_ablate,
+        direction_vectors=direction_vectors,
+        multiplier=multiplier,
+        all_positions=all_positions,
+        extension="pt",
+    )
+    if cached and file.exists():
+        return torch.load(file.path)
+    if layers_to_ablate == "all":
+        layers_to_ablate = list(range(model.cfg.n_layers))
+    experiment_names = [
+        "orig",
+        "ablated",
+    ]
+    # get length of prompts in dataset
+    item_len = len(data_loader.dataset[0]["tokens"])
+    experiment_index = {name: i for i, name in enumerate(experiment_names)}
+    out_shape = (
+        len(experiment_names),
+        data_loader.dataset.num_rows,
+        item_len,
+    )
+    output = torch.zeros(
+        out_shape,
+        device="cpu",
+        dtype=torch.float32,
+    )
+    batch_size = data_loader.batch_size
+    assert batch_size is not None
+    for batch_idx, batch_value in tqdm(enumerate(data_loader), total=len(data_loader)):
+        batch_tokens = batch_value["tokens"].to(device)
+        if all_positions:
+            batch_pos = batch_value["attention_mask"].to(device)
+        else:
+            batch_pos = batch_value["positions"].to(device)
+
+        # Step 1: original metric without hooks
+
+        # get the loss for each token in the batch
+        orig_loss = model(
+            batch_tokens, return_type="loss", prepend_bos=False, loss_per_token=True
+        )
+        assert isinstance(orig_loss, Tensor)
+        # concatenate column of 0s
+        orig_metric = torch.cat(
+            [torch.zeros((orig_loss.shape[0], 1)).to(device), orig_loss], dim=1
+        )
+        output[
+            experiment_index["orig"],
+            batch_idx * batch_size : (batch_idx + 1) *
+            batch_size,
+        ] = orig_metric.cpu()
+
+        # Step 2: repeat with tokens ablated
+
+        for layer in layers_to_ablate:
+            hook = partial(
+                ablation_hook_base,
+                cached_means=cached_means,
+                direction_vectors=direction_vectors,
+                multiplier=multiplier,
+                pos_mask=batch_pos,
+                layer=layer,
+            )
+            model.blocks[layer].hook_resid_post.add_hook(hook)
+
+        # get the loss for each token when run with hooks
+        hooked_loss = model(
+            batch_tokens, return_type="loss", prepend_bos=False, loss_per_token=True
+        )
+        # concatenate column of 0s
+        hooked_loss = torch.cat(
+            [torch.zeros((hooked_loss.shape[0], 1)).to(device), hooked_loss], dim=1
+        )
+        loss_diff = hooked_loss - orig_metric
+        ablated_metric = mask_loss_at_next_positions(
```

Comment:
Should remove this reference to the old masking function and cache all of the loss diffs here